src/connectors/configure-related-items/connectConfigureRelatedItems.ts

```diff
@@ -0,0 +1,155 @@
+     */
+    score: number;
+  };
+};
+
+export interface ConfigureRelatedItemsConnectorParams {
+  /**
+   * The reference hit to extract the filters from.
+   */
+  hit: ResultHit;
+  /**
+   * The schema to create the optional filters.
+   * Each key represents an attribute from the hit.
+   */
+  matchingPatterns: MatchingPatterns;
+  /**
+   * Function called to transform the search parameters generated.
+   */
+  transformSearchParameters?(
+    searchParameters: SearchParameters
+  ): PlainSearchParameters;
+}
+
+type ConfigureConfigureRelatedItemsWidgetFactory<
+  TConfigureRelatedItemsWidgetParams
+> = WidgetFactory<
+  ConfigureRelatedItemsConnectorParams & TConfigureRelatedItemsWidgetParams
+>;
+
+type ConfigureRelatedItemsConnector = <TConfigureRelatedItemsWidgetParams>(
+  render?: ConfigureRenderer<ConfigureConnectorParams>,
+  unmount?: Unmounter
+) => ConfigureConfigureRelatedItemsWidgetFactory<
+  TConfigureRelatedItemsWidgetParams
+>;
+
+const withUsage = createDocumentationMessageGenerator({
+  name: 'configure-related-items',
+  connector: true,
+});
+
+function createOptionalFilter({
+  attributeName,
+  attributeValue,
+  attributeScore,
+}) {
+  return `${attributeName}:${attributeValue}<score=${attributeScore || 1}>`;
+}
+
+const connectConfigureRelatedItems: ConfigureRelatedItemsConnector = (
+  renderFn,
+  unmountFn
+) => {
+  warning(
+    false,
+    'ConfigureRelatedItems is an experimental widget that is subject to change in next minor versions.'
+  );
+
+  return widgetParams => {
+    const { hit, matchingPatterns, transformSearchParameters = x => x } =
+      widgetParams || ({} as typeof widgetParams);
+
+    if (!hit) {
+      throw new Error(withUsage('The `hit` option is required.'));
+    }
+
+    if (!matchingPatterns) {
+      throw new Error(withUsage('The `matchingPatterns` option is required.'));
+    }
+
+    const optionalFilters = Object.keys(matchingPatterns).reduce<
+      Array<string | string[]>
+    >((acc, attributeName) => {
+      const attribute = matchingPatterns[attributeName];
+      const attributeValue = hit[attributeName];
+      const attributeScore = attribute.score;
+
+      if (Array.isArray(attributeValue)) {
+        acc.push(
+          attributeValue.map(attributeSubValue => {
+            return createOptionalFilter({
+              attributeName,
+              attributeValue: attributeSubValue,
+              attributeScore,
+            });
+          })
+        );
+      } else if (typeof attributeValue === 'string') {
+        acc.push(
+          createOptionalFilter({
+            attributeName,
+            attributeValue,
+            attributeScore,
+          })
+        );
+      } else {
+        warning(
+          false,
+          `
+The \`matchingPatterns\` option returned a value of type \`${typeof attributeValue}\` for the "${attributeName}" key. This value was not sent to Algolia because it's not supported by \`optionalFilters\`.
```

Comment:
since both `typeof {} === "object"` and `typeof [] === "object"` this formulation feels a bit strange
```
The `matchingPatterns` option returned a value of type `object` 
```
I'm not sure how to improve it though

Reply:
That's a an option. Note that although generic, it doesn't solve the issue for class instances other than Array. `typeof new Date() === "object"`.
Something using `Object.prototype.toString.call(attributeValue)` would be more reliable.

I just noticed we don't say what the types expected: `this value is expected to be a string or an array`. It is useful information for someone seeing this error.

Reply:
That's a good first step, we can use `Object.prototype.toString.call` which is more accurate.

About the expected types, that's why I link to the documentation. I can add a sentence that makes it more explicit.

Reply:
Adding more explanation about the warning in 168942fc59b76ed35159f77564126c42813a54a6.

@tkrugg I'll send a separate PR that address more correct types in a more global way.

Reply:
PR sent: #4249.

Reply:
Implemented in 6644967387b759548dd953acfc27b29e779cdf9e.